Add Script.signal() to signal a process that it should terminate

Goodwine opened the pull request.

nex3 commented on lib/src/script.dart:
```diff
@@ -153,6 +153,21 @@ class Script {
+  ///
+  /// Returns whether the signal was successfully delivered. Otherwise it
+  /// usually means the process is already dead.
+  FutureOr<bool> signal([ProcessSignal signal = ProcessSignal.sigterm]) async =>
```
User-facing methods should never be typed as `FutureOr`, since it's strictly more annoying to deal with than a guaranteed `Future`.

That said, does this really need to be asynchronous at all? We can define "queued up a signal to be delivered once the `fromComponents` callback completes" as "delivered successfully", since it _will be_ delivered and acted on in the future. I'm worried that returning a future will make callers think that this returns once the process has reacted to the signal, which we can't reliably detect.

Making this return synchronously also avoids us needing to think about whether and when this should throw errors.

Goodwine replied:
I had made it asynchronous on purpose, but you're right that this may send a misleading guarantee so I made it synchronous now, this unfortunately means that for some scripts the signal handler can't be called ASAP and needs some sort of delay for the futures to resolve

nex3 commented on lib/src/script.dart:
```diff
@@ -238,6 +253,10 @@ class Script {
   /// The [name] is a human-readable name for the script, used for debugging and
   /// error messages. It defaults to `"capture"`.
+  ///
+  /// The script can only be interrupted by closing the [Script.stdin] stream.
+  /// Any [signal] is ignored and returns whether or not the script is sill
+  /// running.
```
I think we should give `Script.capture()` callers an optional argument to explicitly handle signals. The expected semantics are substantially different than closing stdin (which may already be a meaningful indicator of just "end of data").

Goodwine replied:
For capturing instead of terminating the stdout (which would not terminate the callback itself) I added an `onSignal` which users can then use if they want cancellation logic in their script

nex3 commented on lib/src/script.dart:
```diff
@@ -153,6 +153,21 @@ class Script {
   final _outputCloser = StreamCloser<List<int>>();
 
+  /// The script's signal handler to terminate the process.
+  late final Future<FutureOr<bool> Function([ProcessSignal])> _signalHandler;
```
Why make the signal argument optional?

Goodwine replied:
fits Process.kill, but I'll make it non optional for the internal type

nex3 reviewed (changes requested):
I think we should add more signal-handling to more types of `Script`. I think `Script.signal()` should mean:

* If the `Script` is a Unix-style OS process, pass the given signal to the process.
* If the `Script` is a thin wrapper over one or more other `Script`s, pass the signal to the children.
* Otherwise, terminate the script if possible regardless of the signal type.

So for example, `ByteStreamExtensions._asScript` should close the stream when signaled (you can use [`StreamCloser`](https://pub.dev/documentation/async/latest/async/StreamCloser-class.html) for this).

Goodwine commented on lib/src/script.dart:
```diff
@@ -153,6 +153,21 @@ class Script {
+  /// Defaults to [ProcessSignal.sigterm].
+  ///
+  /// On platforms without signal support, including Windows, the call ignores
+  /// [signal] and the process terminates in a platform specific way.
```
I have made it so any script currently supported (except those that are "captured") can be terminated via `kill`

nex3 commented on lib/src/extensions/byte_stream.dart:
```diff
@@ -60,19 +60,26 @@ extension ByteStreamExtensions on Stream<List<int>> {
-          exitCodeCompleter.future,
-          ([ProcessSignal? _]) => !exitCodeCompleter.isCompleted);
+          exitCodeCompleter.future, (_) async {
+        if (exitCodeCompleter.isCompleted) return false;
```
We shouldn't ask callers to manually check for this; the `Script` simply shouldn't invoke `onSignal` after it's already emitted an exit code.

nex3 replied:
That's right.

nex3 commented on lib/src/script.dart:
```diff
@@ -154,18 +154,18 @@ class Script {
+  /// Otherwise the signal could not be sent, usually meaning that the process
+  /// is already dead.
+  Future<bool> kill([ProcessSignal signal = ProcessSignal.sigterm]) async =>
       (await _signalHandler)(signal);
```
As I mentioned earlier, this shouldn't invoke the signal handler if the script has already exited.

nex3 commented on test/signal_test.dart:
```diff
@@ -0,0 +1,385 @@
+
+      expect(await script.kill(), true);
+
+      expect(script.done, throwsScriptException(-15));
```
It's probably good to verify that this didn't take 100 seconds to trigger as well.

nex3 replied:
Rather than giving the test a short timeout, which runs the risk of making it flaky, you could start a stopwatch before `script.kill()` and then verify that it's below some threshold after `script.done`. Alternatively, you could have the `mainScript()` wait _forever_, rather than for a set amount of time, for example by a `while (true)` loop with an await inside it.

nex3 commented on test/signal_test.dart:
```diff
@@ -0,0 +1,385 @@
+  group('on Script.capture', () {
+    test("does not interrupt by default", () async {
+      var script = Script.capture((_) async {
+        await Future<void>.delayed(Duration(seconds: 3));
```
Relying on wall-clock sleeps is a recipe for slow tests and flaky failures. It's often better to use an external signal (like a `Completer`) to explicitly control which futures fire when, but in a pure-Dart test like this you can also use the [`fake_async` package](https://pub.dev/packages/fake_async) to use relative timings without relying on the vicissitudes of real time.

nex3 commented on test/signal_test.dart:
```diff
@@ -0,0 +1,385 @@
+    test("can be captured with onSignal", () async {
+      late Script innerScript;
+      var script = Script.capture((_) async {
+        innerScript = _watchSignalsAndExit();
```
It's probably a good idea to avoid using physical subprocesses in these tests when it's not crucial to the behavior being tested, for performance if nothing else.

nex3 commented on test/signal_test.dart:
```diff
@@ -0,0 +1,385 @@
+    script.stdout.lines.listen(events.add, onDone: () => done = true);
+
+    script.stdin.writeln('before');
+    await Future<void>.delayed(Duration(seconds: 1));
```
For uses like this where you're just trying to let the event queue run dry, you can use [`pumpEventQueue()`](https://pub.dev/documentation/test_api/latest/hooks/pumpEventQueue.html).

nex3 reviewed (changes requested):
Additional cases to test:

* `xargs` callbacks shouldn't be invoked once a signal has been received
* prints in a signal handler should be routed through the `Script`
* a signal handler throwing (both `ScriptError` and otherwise)
* the correct signal type is passed into both a physical subprocess and a custom handler
* `Script.kill()` doesn't invoke a custom signal handler after the script has exited

nex3 commented on lib/src/extensions/line_stream.dart:
```diff
@@ -181,22 +180,41 @@ extension LineStreamExtensions on Stream<String> {
-        }
-      }
-    }, name: name ?? 'xargs');
+    var signalCloser = StreamCloser<String>();
```
I'm not saying to do away with `Script.capture()`, I'm saying convert the contents of its callback from an async function with `await for` into an explicit call to `Stream.listen()`. You definitely still want to do that _within_ `Script.capture()`, though.

Goodwine replied:
One very interesting thing, at least for me is how `onDone` is not executed when calling `subscription.cancel()` 😱 so I had to have a `Completer` in there just to pick up the signal

nex3 commented on lib/cli_script.dart:
```diff
@@ -398,23 +397,25 @@ IOSink append(String path) => File(path).openWrite(mode: FileMode.append);
+/// The [callback] can't be interrupted by calling [kill], but the [onSignal]
+/// callback allows capturing those signals so the callback may react
+/// appropriately. Calling [kill] returns `true` even without an [onSignal]
+/// callback because the stream is still interrupted.
```
```suggestion
/// callback if the stream is interrupted.
```

This makes it clearer that killing the script after it's completed will return `false`.

nex3 commented on lib/cli_script.dart:
```diff
@@ -398,23 +397,25 @@ IOSink append(String path) => File(path).openWrite(mode: FileMode.append);
+/// an existing string stream rather than [stdin].
 Script xargs(FutureOr<void> callback(List<String> args),
-    {int? maxArgs, String? name}) {
-  if (maxArgs != null && maxArgs < 1) {
-    throw RangeError.range(maxArgs, 1, null, 'maxArgs');
-  }
```
We should still check this eagerly so that we throw the error synchronously and outside the `Script.capture()` block.

nex3 commented on lib/cli_script.dart:
```diff
@@ -398,23 +397,25 @@ IOSink append(String path) => File(path).openWrite(mode: FileMode.append);
 ///
 /// If [name] is passed, it's used as the name of the spawned script.
 ///
+/// [Script.kill] returns `true` if the stream was interrupted and the script
+/// exits with [Script.exitCode] `143`, or `false` if the stream was already
+/// closed.
```
In these documentation blocks,I'd lead with the primary effect: "[Script.kill] closes the input stream. It returns ..."

nex3 commented on lib/src/script.dart:
```diff
@@ -153,6 +153,28 @@ class Script {
+  /// process. On platforms without signal support, including Windows, the
+  /// [signal] is ignored and the process terminates in a platform specific way.
+  ///
+  /// Returns `true` if the signal is successfully delivered to the process.
```
```suggestion
  /// Returns `true` if the signal is successfully delivered to the script.
```

Not all scripts are processes

nex3 commented on lib/src/script.dart:
```diff
@@ -369,16 +425,21 @@ class Script {
+              Future.wait(list.map((script) => script.exitCode)).then(
+                  (exitCodes) => exitCodes.lastWhere((code) => code != 0,
+                      orElse: () => 0)),
+              (ProcessSignal signal) =>
```
you shouldn't need to add a type annotation here.

nex3 commented on lib/src/script.dart:
```diff
@@ -480,7 +541,7 @@ class Script {
   /// If [silenceStartMessage] is `false` (the default), this prints a message
   /// in debug mode indicating that the script has started running.
   Script._(this.name, StreamConsumer<List<int>> stdin, Stream<List<int>> stdout,
-      Stream<List<int>> stderr, Future<int> exitCode,
+      Stream<List<int>> stderr, Future<ScriptComponents> components,
```
This new signature means you're passing in a bunch of these values twice: once directly and once through `components`. The whole idea of this constructor is to avoid using `ScriptComponents` in the first place—so can it just take `_signalHandler` as a direct argument?

Goodwine replied:
So unfortunately given the way the code is structured I would have to either a) convert to a `factory` constructor which then requires modifying how `BufferedScript` works because it calls `super.fromComponents`, or b) create another holding class for both the exit-code and signal-handler

I would prefer (a) except that it makes inheritance for BufferedScript harder, so I will go with (b)

Goodwine replied:
or... I could use a completer 🤔 yeah that may be nicer

Goodwine commented:
PTAL, I added the requested tests and addressed feedback, the xargs thingy in the LineStream was interesting to deal with `listen` rather than `await for`, and the `Completer` would immediately throw unless I "wait" to release the event from within the `Script.capture` callback 🤔

nex3 commented on lib/src/extensions/line_stream.dart:
```diff
@@ -181,22 +180,38 @@ extension LineStreamExtensions on Stream<String> {
+    late StreamSubscription<void> subscription;
+    return Script.capture(
+        (_) async {
+          var chunks = maxArgs != null ? slices(maxArgs) : toList().asStream();
```
`toList()` here won't actually cancel the underlying stream when the script is killed (something to add a test for!), because it just returns a `Future` which doesn't have any notion of cancelablity.

The more I think about this the more I think you were probably right to use `StreamCloser` in the first place—trying to actually manage an explicit subscription here is probably too complex. Sorry for the churn!

Goodwine replied:
hm I guess I could default to max int 64 🤔

nex3 replied:
This is clever... but I think your original code with `StreamCloser` was probably still cleaner.

nex3 commented on lib/src/extensions/line_stream.dart:
```diff
@@ -181,22 +180,38 @@ extension LineStreamExtensions on Stream<String> {
+    return Script.capture(
+        (_) async {
+          var chunks = maxArgs != null ? slices(maxArgs) : toList().asStream();
+          subscription = chunks.asyncMap(callback).listen((_) {},
+              onDone: () => completer.complete(0), cancelOnError: true);
```
Something to note when working with streams in the future: you should *always* add an `onError` handler. Otherwise, any error that comes through the stream will be treated as totally unhandled and will probably crash the entire application without a way for callers to handle it.

Goodwine replied:
I added a test that correctly captures the exception when the stream gets `addError` even without an `onError` handler, I guess the `Script.capture` took care of it for me here and otherwise it would have crashed as you mention, but I added it nonetheless for completeness

nex3 commented on lib/src/script.dart:
```diff
@@ -479,8 +543,13 @@ class Script {
+      StreamConsumer<List<int>> stdin,
+      Stream<List<int>> stdout,
+      Stream<List<int>> stderr,
+      Future<bool Function(ProcessSignal)> signalHandler,
```
I think passing this in as a future is wrong, for a few reasons:

1. This being a future means that there's a gap between creating a `Script` and it fully initializing when `Script.kill()` won't do anything.

2. It makes sense to be able to kill a `Script` while it's initializing. For example, in `Script()`, there's an async gap when evaluating the script's arguments—if the script is killed then, it should avoid spawning the subprocess entirely.

3. Making this a non-`Future` avoids a lot of complexity around around completers and a lot of cases where you've otherwise needed to change `Script._()` calls to `Script.fromCompleter()`.

nex3 commented on test/signal_test.dart:
```diff
@@ -0,0 +1,428 @@
+    expect(script.kill(), false);
+  });
+
+  test("interrupts a ByteStream pipe", () async {
```
Any of these tests with streams should also verify that the underlying stream is canceled.

nex3 reviewed (changes requested):
Just a few more changes and then I think we'll be ready to go!

Goodwine commented on lib/src/script.dart:
```diff
@@ -287,36 +337,57 @@ class Script {
 
-    return Script._(scriptName, stdinController.sink, stdoutGroup.stream,
-        stderrGroup.stream, exitCodeCompleter.future);
+    return Script.fromComponents(
+        scriptName,
+        () => ScriptComponents(stdinController.sink, stdoutGroup.stream,
+            stderrGroup.stream, exitCodeCompleter.future), onSignal: (signal) {
```
Interestingly the bytestream pipe starts behaving erratically when I change this one, but everything else works fine if I change the rest, I'll keep looking at why is that

Goodwine replied:
ok so it turns out there is a bug where script.capture halts if the input stream is closed, but only on tests not on real code.. which is very confusing :|

https://github.com/google/dart_cli_script/issues/46

I added a workaround

Goodwine commented on lib/src/script.dart:
```diff
@@ -192,15 +216,31 @@ class Script {
-          process.stdin, process.stdout, process.stderr, process.exitCode);
+          process!.stdin, process!.stdout, process!.stderr, process!.exitCode);
+    }, (signal) {
+      capturedSignal = null;
```
My train of thought was:
1. User calls `kill` before process is instantiated, the signal is captured
2. User calls `kill` after process is instantiated but before the program runs over line 235, the signal isn't captured and the program calls `process.kill` from line 241, then the program resumes line 235 and it triggers `process.kill` (unexpectedly) again

but for that to happen the user would have had to call `process.kill` twice.. so it makes sense the program calls `process.kill` twice 🤔 

So I guess removing is fine? That said now I wonder.. if a user calls `script.kill` N times before the process is initialized we will only send out the last signal, but I guess that's fine too

Goodwine commented:
I have updated the branch to remove the workaround now that the script.capture bug is fixed, thanks!

nex3 reviewed (approved):
Thanks for bearing with me on this one :smile:. I know async code is a huge hassle, but it's also very important to get right, since the failure states (as you've seen) are fiendishly difficult to debug.
